src/main/java/com/conveyal/r5/analyst/scenario/RemoveStops.java

```diff
@@ -147,23 +158,63 @@ public TripPattern processTripPattern (TripPattern originalTripPattern) {
             int accumulatedRideTime = 0;
             int prevInputDeparture = 0;
             int prevOutputDeparture = 0;
+            int nStopsRemovedSinceLastStop = 0;
             for (int i = 0, j = 0; i < removeStop.length; i++) {
                 int rideTime = originalSchedule.arrivals[i] - prevInputDeparture;
                 int dwellTime = originalSchedule.departures[i] - originalSchedule.arrivals[i];
                 prevInputDeparture = originalSchedule.departures[i];
                 if (removeStop[i]) {
-                    // The current stop will not be included in the output. Record the travel time.
+                    // The current stop will not be included in the output. Record the travel time and subtract off the
+                    // seconds saved. This may cause accumulatedRideTime to go negative, which may be fine since we're putting
+                    // all the time savings before the stop. After we've accumulated ride time for a whole segment,
+                    // we check to make sure it's positive.
                     accumulatedRideTime += rideTime;
+                    nStopsRemovedSinceLastStop++;
                     if (j == 0) {
                         // Only accumulate dwell time at the beginning of the output trip, to preserve the time offset of
                         // the first arrival in the output. After that, dwells are not retained for skipped stops.
                         accumulatedRideTime += dwellTime;
                     }
                 } else {
-                    newSchedule.arrivals[j] = prevOutputDeparture + accumulatedRideTime + rideTime;
+                    accumulatedRideTime += rideTime;
+                    int secondsToRemove = secondsSavedAtEachStop * nStopsRemovedSinceLastStop;
+                    if (nStopsRemovedSinceLastStop > 0 && accumulatedRideTime < secondsToRemove) {
+                        // figure out which stops are causing the problem
+                        int[] problemStops = new int[nStopsRemovedSinceLastStop];
+                        for (int removedStop = i - 1, index = nStopsRemovedSinceLastStop - 1; removedStop >= i - nStopsRemovedSinceLastStop; removedStop--, index--) {
+                            problemStops[index] = originalTripPattern.stops[removedStop];
+                        }
+
+                        // map them back to names
+                        String problemStopNames = IntStream.of(problemStops)
```

Comment:
Having the error handling code inserted in the conditional body inside the loop stretches out the blocks and makes it a bit harder to see the overall flow. It seems sufficient to just record all the offending stops in a set and handle the error generation outside the inner blocks, potentially even in a function.